src/app/reporting/LoggingManagement.cpp

```diff
@@ -0,0 +1,1306 @@
+    }
+
+    SuccessOrExit(eventDataElementBuilder.GetError());
+
+    err = apContext->mWriter.StartContainer(ContextTag(chip::app::EventDataElement::kCsTag_Data), chip::TLV::kTLVType_Structure,
+                                            dataContainerType);
+    SuccessOrExit(err);
+    // Callback to write the EventData
+    err = aEventWriter(apContext->mWriter, apAppData);
+    SuccessOrExit(err);
+
+    err = apContext->mWriter.EndContainer(dataContainerType);
+    SuccessOrExit(err);
+
+    eventDataElementBuilder.EndOfEventDataElement();
+    SuccessOrExit(eventDataElementBuilder.GetError());
+
+    err = apContext->mWriter.Finalize();
+    SuccessOrExit(err);
+
+    // only update mFirst if an event was successfully written.
+    if (apContext->mFirst)
+    {
+        apContext->mFirst = false;
+    }
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        apContext->mWriter = checkpoint;
+    }
+    else
+    {
+        // update these variables since BlitEvent can be used to track the
+        // state of a set of events over multiple calls.
+        apContext->mCurrentEventNumber++;
+        {
+            apContext->mCurrentTime = apOptions->timestamp.systemTimestamp;
+        }
+    }
+    return err;
+}
+
+/**
+ * @brief Helper function to skip writing an event corresponding to an allocated
+ *   event id.
+ *
+ * @param[inout] apContext   EventLoadOutContext, initialized with stateful
+ *                          information for the buffer. State is updated
+ *                          and preserved by BlitEvent using this context.
+ *
+ */
+void LoggingManagement::SkipEvent(EventLoadOutContext * apContext)
+{
+    apContext->mCurrentEventNumber++; // Advance the event id without writing anything
+}
+
+/**
+ * @brief Create LoggingManagement object and initialize the logging management
+ *   subsystem with provided resources.
+ *
+ * Initialize the LoggingManagement with an array of LogStorageResources.  The
+ * array must provide a resource for each valid priority level, the elements
+ * of the array must be in increasing numerical value of priority (and in
+ * decreasing priority); the first element in the array corresponds to the
+ * resources allocated for the most critical events, and the last element
+ * corresponds to the least important events.
+ *
+ * @param[in] inMgr         ExchangeManager to be used with this logging subsystem
+ *
+ * @param[in] inNumBuffers  Number of elements in inLogStorageResources array
+ *
+ * @param[in] inLogStorageResources  An array of LogStorageResources for each priority level.
+ *
+ * @note This function must be called prior to the logging being used.
+ */
+void LoggingManagement::CreateLoggingManagement(Messaging::ExchangeManager * apMgr, size_t aNumBuffers,
+                                                const LogStorageResources * const apLogStorageResources)
+{
+    new (&sInstance) LoggingManagement(apMgr, aNumBuffers, apLogStorageResources);
```

Comment:
Why is this placement new required?
Could we have setters instead?

Reply:
This pattern is not correct in general. sInstance is already initialized by any conforming C++ implementation by the time this function executes.

If LoggingManagement is trivially destructible, it may be OK. In that case we should `static_assert(std::is_trivially_destructible<LoggingManagement>::value, "LoggingManagement must be trivially destructible");`

I don't think that assert is true, because it contains a counter member.

Reply:
Just confirm it is trivially_destructible via printing log.

Reply:
If it is, add a static assert to that effect, please.

Reply:
Done, thanks